skglm/solvers/gram_cd.py

```diff
@@ -0,0 +1,117 @@
+import numpy as np
+from numba import njit
+from scipy.sparse import issparse
+from skglm.utils import AndersonAcceleration
+
+
+def gram_cd_solver(X, y, penalty, max_iter=20, w_init=None,
+                   use_acc=True, cd_type='greedy', tol=1e-4, verbose=False):
+    """Run coordinate descent while keeping the gradients up-to-date with Gram updates.
+
+    Minimize::
+        1 / (2*n_samples) * norm(y - Xw)**2 + penalty(w)
+
+    Which can be rewritten as::
+        w.T @ Q @ w / (2*n_samples) - q.T @ w / n_samples + penalty(w)
+
+    where::
+        Q = X.T @ X (gram matrix)
+        q = X.T @ y
+    """
+    n_samples, n_features = X.shape
+    scaled_gram = X.T @ X / n_samples
+    scaled_Xty = X.T @ y / n_samples
+    scaled_y_norm2 = np.linalg.norm(y)**2 / (2*n_samples)
+
+    if issparse(X):
+        scaled_gram = scaled_gram.toarray()
+
+    all_features = np.arange(n_features)
+    stop_crit = np.inf  # prevent ref before assign
+    p_objs_out = []
+
+    w = np.zeros(n_features) if w_init is None else w_init
+    scaled_gram_w = np.zeros(n_features) if w_init is None else scaled_gram @ w_init
+    opt = penalty.subdiff_distance(w, -scaled_Xty, all_features)  # initial: grad = -Xty
```

Comment:
beware that this will not hold if a non zero `w_init` is passed

Reply:
That's a good catch indeed!